vmdb/app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -99,4 +99,27 @@ def store_ids_for_new_records(records, hashes, keys)
       h[:id] = r.id
     end
   end
+
+  # most of the refresh_multi calls follow the same pattern
+  # this pulls it out
+  def save_inventory_assoc(type, parent, hashes, target, find_key, child_keys = [], extra_keys = [])
+    reflection = parent.class.reflect_on_association(type)
+    klass = reflection.class_name.constantize
+    deletes = relation_values(parent, reflection, target)
+
+    save_inventory_multi(type, klass, parent, hashes, deletes, find_key, child_keys, extra_keys)
+    store_ids_for_new_records(parent.send(type), hashes, find_key)
+  end
+
+  # We need to determine our intent:
+  # - make a complete refresh. Delete missing values.
+  # - make a partial refresh. Don't delete missing keys
+  # This generates the "deletes" values based upon this intent
+  def relation_values(parent, reflection, target)
+    # always want to refresh this association
+    reflection = parent.class.reflect_on_association(reflection) if reflection.kind_of?(Symbol)
+    top_level = reflection.options[:dependent] == :destroy
```

Comment:
Some associations (provision_manager -> operating_system_flavor) are the definitive source of the association. We want to delete those base records if they are missing. So we declare `:dependent => :destroy`.

Other associations (configuration_profile => :operating_system_flavor) are references, but we don't want to delete the flavor from the db if it is removed. So we don't declare `:dependent => :destroy`.

The same logic that goes for active record goes for the refresh.

I called it a `top_level` association - it owns the lifespan of the child objects.

In our code we tend to just hardcode the logic into the ruby. But I thought we could just leverage AR.

I'm pretty sure this is universal in our refreshes, but we can make sure when/if we actually integrate this.